File: LeetCodeRecord/AC 1.0/139-WordBreak.java
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

class WordBreak {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<String>(wordDict);
        int maxStrLen= 0;
        boolean[] dp = new boolean[s.length() + 1];
        
        // 记录最长字符串的长度
        for (String str : wordSet) {
            if (str.length() > maxStrLen) {
                maxStrLen = str.length();
            }
        }

        dp[0] = true;
        for (int i = 1; i <= s.length(); i++) {
            for (int j = i-1; j >= 0 && (i - j) <= maxStrLen; j--) {
                dp[i] = dp[j] && wordSet.contains(s.substring(j, i));
                if (dp[i]) {
                    break;
                }
            }
        }

        return dp[s.length()];
    }

    public static void main(String[] args) {
        WordBreak wb = new WordBreak();
        List<String> list = new LinkedList<String>();
        list.add("cats");
        list.add("dog");
        list.add("sand");
        list.add("and");
        list.add("cat");
        System.out.println(wb.wordBreak("catsandog", list));
        // System.out.println("leetcode".substring(0, 1));
    }
}

File: Top200/139-WordBreak.go
package main

func wordBreak(s string, wordDict []string) bool {
	sLen := len(s)

	dp := make([]bool, sLen+1)
	dp[0] = true

	for i := 1; i <= sLen; i++ {
		for j := 0; j < len(wordDict); j++ {
			l := len(wordDict[j])
			if l > i {
				continue
			}

			if wordDict[j] == s[i-l:i] {
				dp[i] = (true && dp[i-l]) || dp[i]
			}
		}
	}

	return dp[sLen]
}
